tracer/src/Datadog.Trace/Telemetry/TelemetryController.cs

```diff
@@ -0,0 +1,245 @@
+        }
+
+        private void CurrentDomain_OnAssemblyLoad(object sender, AssemblyLoadEventArgs e)
+        {
+            RecordAssembly(e.LoadedAssembly);
+        }
+
+        private void RecordAssembly(Assembly assembly)
+        {
+            try
+            {
+                _dependencies.AssemblyLoaded(assembly.GetName());
+            }
+            catch (Exception ex)
+            {
+                Log.Warning(ex, "Error recording loaded assembly");
+            }
+        }
+
+        private async Task PushTelemetryLoopAsync()
+        {
+#if !NET5_0_OR_GREATER
+            var tasks = new Task[2];
+            tasks[0] = _tracerInitialized.Task;
+#endif
+            while (true)
+            {
+#if !NET5_0_OR_GREATER
+                if (_tracerInitialized.Task.IsCompleted)
+                {
+                    tasks[0] = _processExit.Task;
+                }
+#endif
+
+                if (_processExit.Task.IsCompleted)
+                {
+                    Log.Debug("Process exit requested, ending telemetry loop");
+                    var sendAppClosingTelemetry = _processExit.Task.Result;
+                    if (sendAppClosingTelemetry)
+                    {
+                        Log.Debug("Process exit requested, ending telemetry loop");
+                        await PushAppClosingTelemetry().ConfigureAwait(false);
+                    }
+
+                    return;
+                }
+
+                await PushTelemetry().ConfigureAwait(false);
+
+#if NET5_0_OR_GREATER
+                // .NET 5.0 has an explicit overload for this
+                await Task.WhenAny(
+                               Task.Delay(_bucketDuration),
+                               _tracerInitialized.Task.IsCompleted
+                                    ? _processExit.Task
+                                    : _tracerInitialized.Task)
+                          .ConfigureAwait(false);
+#else
+                tasks[1] = Task.Delay(_sendFrequency);
+                await Task.WhenAny(tasks).ConfigureAwait(false);
+#endif
+            }
+        }
+
+        private async Task PushTelemetry()
+        {
+            try
+            {
+                var application = _configuration.GetApplicationData();
+                var host = _configuration.GetHostData();
+                if (application is null || host is null)
+                {
+                    Log.Debug("Telemetry not initialized, skipping");
+                    return;
+                }
+
+                // These calls change the state of the collectors, so must use the data
+                var success = await PushTelemetry(application, host, sendHeartbeat: true).ConfigureAwait(false);
+                if (!success)
+                {
+                    FatalError = true;
+                    Log.Debug("Unable to send telemetry, ending telemetry loop");
+                    TerminateLoop(sendAppClosingTelemetry: false);
```

Comment:
We're currently stopping immediately _only_ if we get a 404 from the endpoint _or_ an exception (e.g. if a connection can't be made). Those aren't likely to be resolved, especially as we require the most recent version of the agent to be deployed, so the most likely situation is unrecoverable failure. So I think terminating in those cases is the best approach.

That said, we're currently not retrying for other errors too. My original thinking was that this wasn't necessary, as we'll naturally send an update on the next cycle, but that's not necessarily the case, as subsequent pings are likely to be heartbeats. 

I think we should do one or both of the following:
* Enable retrys for transient errors
* Rework the algorithm, so that if it fails to send the `app-started` event, it attempts to send it on subsequent loops

Reply:
The second solution is appealing. But looking back at this after 2 weeks, I wonder if it's that important. I mean we had no telemetry at all, and missing a few events won't be an issue I assume. So as best is enemy of the good, maybe we shouldn't do anything here. At least in this PR anyway

Reply:
😄 I guess I should have checked my notifications before implementing the second option in https://github.com/DataDog/dd-trace-dotnet/pull/2241/commits/d042be18a278e54b61934c01324bd0a62ab4c6e1

Also considered the fact that we can read the `/info` endpoint from the endpoint, to determine whether to send telemetry, but that seemed overly complex, especially as we _may_ want to send directly to the intake (e.g. for testing etc, in integration tests). Just mentioning it here for completeness (and we _could_ add it later if we deem it useful)